src/scenechange/mod.rs

```diff
@@ -40,34 +40,46 @@ impl<T: Pixel> Default for SceneChangeDetector<T> {
   }
 }
 
+pub fn is_changed<T: Pixel>(prev: &Arc<Frame<T>>, curr: &Arc<Frame<T>>, threshold: u8) -> bool {
+  let len = curr.planes[0].cfg.width * curr.planes[0].cfg.height;
+  let delta_yuv = prev
+    .iter()
+    .zip(curr.iter())
+    .map(|(last, cur)| {
+      (
+        (i16::cast_from(cur.0) - i16::cast_from(last.0)).abs() as u64,
+        (i16::cast_from(cur.1) - i16::cast_from(last.1)).abs() as u64,
+        (i16::cast_from(cur.2) - i16::cast_from(last.2)).abs() as u64
+      )
+    })
+    .fold((0, 0, 0), |(ht, st, vt), (h, s, v)| (ht + h, st + s, vt + v));
+  let delta_yuv = (
+    (delta_yuv.0 / len as u64) as u16,
+    (delta_yuv.1 / len as u64) as u16,
+    (delta_yuv.2 / len as u64) as u16
+  );
+  let delta_avg = ((delta_yuv.0 + delta_yuv.1 + delta_yuv.2) / 3) as u8;
+  delta_avg >= threshold
+}
+
 impl<T: Pixel> SceneChangeDetector<T> {
   pub fn new(bit_depth: usize) -> Self {
     let mut detector = Self::default();
     detector.threshold = detector.threshold * bit_depth as u8 / 8;
     detector
   }
 
+  pub fn set_last_frame(&mut self, ref_frame: Arc<Frame<T>>, frame_num: usize) {
+    self.last_frame = Some((frame_num, ref_frame));
+  }
+
   pub fn detect_scene_change(&mut self, curr_frame: Arc<Frame<T>>, frame_num: usize) -> bool {
     let mut is_change = false;
 
-    match self.last_frame {
-      Some((last_num, ref last_frame)) if last_num == frame_num - 1 => {
-        let len = curr_frame.planes[0].cfg.width * curr_frame.planes[0].cfg.height;
-        let delta_yuv = last_frame.iter().zip(curr_frame.iter())
-          .map(|(last, cur)| (
-            (i16::cast_from(cur.0) - i16::cast_from(last.0)).abs() as u64,
-            (i16::cast_from(cur.1) - i16::cast_from(last.1)).abs() as u64,
-            (i16::cast_from(cur.2) - i16::cast_from(last.2)).abs() as u64
-          )).fold((0, 0, 0), |(ht, st, vt), (h, s, v)| (ht + h, st + s, vt + v));
-        let delta_yuv = (
-          (delta_yuv.0 / len as u64) as u16,
-          (delta_yuv.1 / len as u64) as u16,
-          (delta_yuv.2 / len as u64) as u16
-        );
-        let delta_avg = ((delta_yuv.0 + delta_yuv.1 + delta_yuv.2) / 3) as u8;
-        is_change = delta_avg >= self.threshold;
+    if let Some((last_num, ref last_frame)) = self.last_frame {
+      if last_num == frame_num - 1 {
+        is_change = is_changed(last_frame, &curr_frame, self.threshold);
```

Comment:
Extracting this part into a function in my opinion does not do much, as it reduces this function to only a half dozen lines that defer the processing elsewhere. Is there a constraint that requires `is_changed()` to be in its own function?

If kept, I would suggest to change the name to a clearer one with context of what is "changed", as without reading it in context, it is hard to tell what it means at a glance.

Reply:
It is a step in between some larger rework. I can rename it to something different or drop this patch for the time being. Ideally I could do in a different way.

Reply:
If it is needed for later work then it is fine to merge, I would just like to know the purpose of factoring it out more specifically.

Reply:
Patch dropped from this set for now.